src/TypeShape.Examples/MsgPackSerializer/MsgPackSerializer.cs

```diff
@@ -0,0 +1,282 @@
+using MessagePack;
+using System.Buffers;
+using System.Diagnostics.CodeAnalysis;
+using System.Text;
+using TypeShape.Abstractions;
+using TypeShape.Examples.Utilities;
+
+namespace TypeShape.Examples.MsgPackSerializer;
+
+/// <summary>
+/// A msgpack serializer built on top of TypeShape.
+/// </summary>
+public static class MsgPackSerializer
+{
+    /// <summary>
+    /// Serializes a value.
+    /// </summary>
+    /// <param name="writer">The writer that receives the msgpack bytes.</param>
+    /// <param name="value">The value to be serialized.</param>
+    /// <typeparam name="T">The type of value to be serialized.</typeparam>
+    public static void Serialize<T>(IBufferWriter<byte> writer, in T? value)
+        where T : IShapeable<T>
+    {
+        MessagePackWriter msgpackWriter = new(writer);
+        SerializerCache<T>.Serializer(ref msgpackWriter, value);
+        msgpackWriter.Flush();
+    }
+
+    /// <summary>
+    /// Deserializes msgpack into a value of a particular type.
+    /// </summary>
+    /// <param name="sequence">The msgpack to deserialize.</param>
+    /// <typeparam name="T">The type of value to deserialize.</typeparam>
+    /// <returns>The deserialized value.</returns>
+    public static T? Deserialize<T>(ReadOnlySequence<byte> sequence)
+        where T : IShapeable<T>
+    {
+        MessagePackReader reader = new(sequence);
+        return SerializerCache<T>.Deserializer(ref reader);
+    }
+
+    private delegate T? Reader<T>(ref MessagePackReader reader);
+
+    private delegate void Writer<T>(ref MessagePackWriter writer, T value);
+
+    private static class SerializerCache<T>
+        where T : IShapeable<T>
+    {
+        private static Writer<T?>? serializer;
+
+        private static Reader<T?>? deserializer;
+
+        internal static Writer<T?> Serializer => serializer ??= SerializeVisitor.GetWriter(T.GetShape());
+
+        internal static Reader<T?> Deserializer => deserializer ??= DeserializeVisitor.GetReader(T.GetShape());
+    }
+
+    private sealed class SerializeVisitor : TypeShapeVisitor
+    {
+        internal static readonly SerializeVisitor Instance = new();
+
+        private static readonly TypeDictionary formatters = new()
+        {
+            { typeof(string), new Writer<string>((ref MessagePackWriter writer, string value) => writer.Write(value)) },
+            { typeof(int), new Writer<int>((ref MessagePackWriter writer, int value) => writer.Write(value)) }
+        };
+
+        internal static Writer<T?> GetWriter<T>(ITypeShape<T> typeShape)
+        {
+            return formatters.GetOrAdd<Writer<T?>>(typeShape, Instance, box => (ref MessagePackWriter writer, T? value) => box.Result(ref writer, value));
+        }
+
+        private SerializeVisitor() { }
+
+        public override object? VisitObject<T>(IObjectTypeShape<T> objectShape, object? state = null)
+        {
+            Dictionary<ReadOnlyMemory<byte>, Writer<T>> properties = new();
```

Comment:
Given that you're not looking this dictionary up, it might make sense to change this to `(byte[], Writer<T>)[]` for faster iteration times when serializing.